lib/Standalone/ToBlockLayoutAndBack.cpp

```diff
@@ -209,6 +167,102 @@ static LogicalResult BlockOpPreconditions(linalg::LinalgOp linalgOp) {
+template <typename CONVOP> static bool hasStride(CONVOP convOp) {
+  if (DenseIntElementsAttr strides = convOp.getStrides()) {
+    auto values = strides.getValues<APInt>();
+    if (llvm::any_of(values, [](const APInt &value) {
+          return value.getSExtValue() != 1;
+        })) {
+      return true;
+    }
+  }
+  return false;
+}
+
+//===----------------------------------------------------------------------===//
+// Conv2DNhwcHwcfOp
+//===----------------------------------------------------------------------===//
+// Original layout: [N][P][Q][K] += [N][H][W][C] * [R][S][C][K]
+// New      layout: [N][K'][P][Q][k] += [N][C'][H][W][c] * [K'][C'][R][S][c][k]
+FailureOr<linalg::GenericOp>
+mlir::linalgx::packConv2DNhwcHwcfOp(RewriterBase &rewriter,
+                                    linalg::Conv2DNhwcHwcfOp convOp,
+                                    ArrayRef<OpFoldResult> tiles) {
+  if (tiles.size() != 2)
+    return rewriter.notifyMatchFailure(convOp, "require 2 tile factors");
+
+  if (convOp.hasDynamicShape())
+    return rewriter.notifyMatchFailure(convOp, "require static shape");
+
+  if (convOp.hasBufferSemantics())
+    return rewriter.notifyMatchFailure(convOp, "require tensor semantics");
+
+  // TODO: (lorenzo) add the stride when you compose the affine map.
+  if (hasStride(convOp))
+    return rewriter.notifyMatchFailure(convOp, "require unit stride");
+
+  Location loc = convOp.getLoc();
+  MLIRContext *ctx = convOp.getContext();
+  SmallVector<Value, 2> reshapedInputTensors;
+
+  SmallVector<Value> inputOperands = convOp.getInputOperands();
+  SmallVector<Value> outputOperands = convOp.getOutputOperands();
+
+  // pack the image and the filter.
+  Value image = inputOperands[0];
+  Value packedImage = toPackLayoutNKPQk(loc, image, tiles[0], rewriter);
+
+  Value filter = inputOperands[1];
+  Value packedFilter = toPackLayoutRSCK_KCRSck(loc, filter, tiles, rewriter);
+  SmallVector<Value> packedInputs = {packedImage, packedFilter};
+
+  // pack the output.
+  Value output = outputOperands[0];
+  Value packedOutput = toPackLayoutNKPQk(loc, output, tiles[0], rewriter);
+
+  // Swap convolution with generic.
+  //         N   K   P   Q   k   C   R   S   c
+  AffineExpr p1, p2, p3, p4, p5, r1, r2, r3, r4;
+  bindDims(ctx, p1, p2, p3, p4, p5, r1, r2, r3, r4);
+  AffineMap mapOut =
+      AffineMap::get(/*dims=*/9, /*symbols=*/0, {p1, p2, p3, p4, p5}, ctx);
+  AffineMap mapImg = AffineMap::get(/*dims=*/9, /*symbols=*/0,
+                                    {p1, r1, p3 + r2, p4 + r3, r4}, ctx);
+  AffineMap mapFil =
+      AffineMap::get(/*dims=*/9, /*symbols=*/0, {p2, r1, r2, r3, r4, p5}, ctx);
+  linalg::GenericOp replacementOp = rewriter.create<linalg::GenericOp>(
+      loc, packedOutput.getType(), packedInputs, ValueRange{packedOutput},
+      ArrayRef<AffineMap>{mapImg, mapFil, mapOut},
+      ArrayRef<StringRef>{
+          getParallelIteratorTypeName(), getParallelIteratorTypeName(),
```

Comment:
this could be cached, at least for code simplicity?

Reply:
Yes, I will clean up tomorrow.